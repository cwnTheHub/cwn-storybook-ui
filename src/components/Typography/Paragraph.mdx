import {
  Canvas,
  Meta,
  Title,
  Subtitle,
  Description,
  Primary,
  Controls,
  Stories,
  Story,
} from "@storybook/blocks";

import * as ParagraphStories from "./Paragraph.stories";

<Meta of={ParagraphStories} />
<Title />

`import {Paragraph} from '@nds_core/nds'`

Block text as an HTML `<p>` element.

### PROPS & METHODS

<Controls />

### Minimal usage

<Canvas of={ParagraphStories.MinimalUsage} />

### Usage criteria

- All body text should be contained in a `Paragraph`, regardless of length.
- If the `Paragraph` is on a dark background, the `invert` prop must be used to maintain sufficient colour contrast.
- If other modifications must be made to the text inside a `Paragraph` such as size or font weight, use the `Text`, `Small`, and `Strong` components.

### Placing text on dark colours

Use the invert prop.

<Story of={ParagraphStories.Inverted} />
