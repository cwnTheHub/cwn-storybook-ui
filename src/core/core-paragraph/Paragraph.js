import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { DependentIconSizeContext, safeRest } from "../../util-helpers";
import * as typography from "../../shared-typography/typography";
import { spacing } from "../../shared-styles";

const paragraphColor = ({ invert }) =>
  invert ? typography.invertedColor : typography.color;
const paragraphInheritColor = ({ inheritColor }) =>
  inheritColor ? { color: "inherit" } : undefined;

const paragraphSize = ({ size }) => typography[size];
const paragraphBold = ({ bold, size }) =>
  bold ? typography.boldFont : typography[`${size}Font`];
const paragraphAlign = ({ align }) => ({
  textAlign: align,
});

export const StyledParagraph = styled.p(
  paragraphColor,
  typography.wordBreak,
  spacing.noSpacing,
  paragraphInheritColor,
  paragraphSize,
  paragraphBold,
  paragraphAlign,
  { sup: typography.sup }
);

const Paragraph = ({ size, invert, children, ...rest }, context) => {
  return (
    <DependentIconSizeContext.Provider value={{ paragraphSize: size, invert }}>
      <StyledParagraph
        {...safeRest(rest)}
        size={size}
        invert={invert}
        inheritColor={context.inheritColor}
      >
        {children}
      </StyledParagraph>
    </DependentIconSizeContext.Provider>
  );
};

Paragraph.propTypes = {
  bold: PropTypes.bool,
  size: PropTypes.oneOf(["small", "medium", "large"]),
  align: PropTypes.oneOf(["left", "center", "right"]),
  invert: PropTypes.bool,
  children: PropTypes.node.isRequired,
};

Paragraph.defaultProps = {
  bold: false,
  size: "medium",
  align: "left",
  invert: false,
};

Paragraph.contextTypes = {
  inheritColor: PropTypes.bool,
};

export default Paragraph;
