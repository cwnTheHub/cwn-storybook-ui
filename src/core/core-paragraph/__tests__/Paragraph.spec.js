import React from "react";
import { shallow, render } from "enzyme";

import * as typography from "../../../shared-typography/typography";

import Paragraph, { StyledParagraph } from "../Paragraph";

describe("Paragraph", () => {
  const doShallow = (props) =>
    shallow(<Paragraph {...props}>Some content</Paragraph>);
  const doRender = (props) =>
    render(<Paragraph {...props}>Some content</Paragraph>);

  it("renders", () => {
    const paragraph = doRender();

    expect(paragraph).toMatchSnapshot();
  });

  /* describe("StyledParagraph", () => {
    const doShallowStyled = (props) =>
      shallow(
        <StyledParagraph size="medium" align="left" {...props}>
          Some content
        </StyledParagraph>
      );
    it("can be bold", () => {
      let paragraph = doShallowStyled({ bold: false });
      expect(paragraph).not.toHaveStyle("font-weight", "700");
      expect(paragraph).toHaveStyle("font-weight", "400");

      paragraph = doShallowStyled({ bold: true });
      expect(paragraph).toHaveStyle("font-weight", "700");
      expect(paragraph).not.toHaveStyle("font-weight", "500");
    });

    it("can be inverted", () => {
      let paragraph = doShallowStyled();
      expect(paragraph).toHaveStyle("color", typography.color.color);
      expect(paragraph).not.toHaveStyle(
        "color",
        typography.invertedColor.color
      );

      paragraph = doShallowStyled({ invert: true });
      expect(paragraph).not.toHaveStyle("color", typography.color.color);
      expect(paragraph).toHaveStyle(
        "color",
        typography.invertedColor.color
      );
    });

    it("can be aligned", () => {
      let paragraph = doShallowStyled();
      expect(paragraph).toHaveStyle("text-align", "left");

      paragraph = doShallowStyled({ align: "right" });
      expect(paragraph).toHaveStyle("text-align", "right");
    });

    it("can be sized", () => {
      let paragraph = doShallowStyled();
      expect(paragraph).toHaveStyle("font-size", "1rem");
      expect(paragraph).toHaveStyle("letter-spacing", "-0.8px");
      expect(paragraph).toHaveStyle("line-height", "1.5");

      paragraph = doShallowStyled({ size: "small" });
      expect(paragraph).toHaveStyle("font-size", "0.875rem");
      expect(paragraph).toHaveStyle("letter-spacing", "-0.6px");
      expect(paragraph).toHaveStyle("line-height", "1.42857");
    });
  }); */

  it("passes additional attributes to the p element", () => {
    const paragraph = doShallow({ id: "the-paragraph" });

    expect(paragraph.find("Paragraph__StyledParagraph")).toBeTruthy();
  });

  it("does not allow custom CSS", () => {
    const paragraph = doShallow({
      className: "my-custom-class",
      style: { color: "hotpink" },
    });

    expect(paragraph).not.toHaveProp("className", "my-custom-class");
    expect(paragraph).not.toHaveProp("style");
  });
});
