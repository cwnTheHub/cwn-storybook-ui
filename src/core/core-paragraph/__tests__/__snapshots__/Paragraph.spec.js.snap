// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph renders 1`] = `
<p
  class="sc-aXZVg evBDiw"
>
  Some content
</p>
`;
